src/reactor.cc

```diff
@@ -0,0 +1,406 @@
+}
+
+void Reactor::CombineInventory(
+    cyclus::toolkit::ResBuf<cyclus::Material>& inventory) {
+  if (!inventory.empty()) {
+    cyclus::Material::Ptr base = inventory.Pop();
+    int count = inventory.count();
+    for (int i = 0; i < count; i++) {
+      cyclus::Material::Ptr m = inventory.Pop();
+      base->Absorb(m);
+    }
+
+    inventory.Push(base);
+  }
+}
+
+void Reactor::ExtractHelium(
+    cyclus::toolkit::ResBuf<cyclus::Material>& inventory) {
+  if (!inventory.empty()) {
+    cyclus::Material::Ptr mat = inventory.Pop();
+    cyclus::CompMap c = mat->comp()->atom();
+    cyclus::compmath::Normalize(&c, mat->quantity());
+
+    cyclus::CompMap He3 = {{20030000, 1}};
+
+    // A threshold of 1e-5 was set to allow tritium_reserve inventories up to
+    // 1000kg. A 1 decade lower threshold prevents tritium_reserve inventories
+    // above 33kg.
+    cyclus::Material::Ptr helium = mat->ExtractComp(
+        c[20030000], cyclus::Composition::CreateFromAtom(He3), 1e-5);
+
+    helium_storage.Push(helium);
+    inventory.Push(mat);
+  }
+}
+
+void Reactor::RecordEvent(std::string name, std::string val) {
+  context()
+      ->NewDatum("ReactorEvents")
+      ->AddVal("AgentId", id())
+      ->AddVal("Time", context()->time())
+      ->AddVal("Event", name)
+      ->AddVal("Value", val)
+      ->Record();
+}
+
+void Reactor::RecordOperationalInfo(std::string name, std::string val) {
+  context()
+      ->NewDatum("ReactorOperationsLog")
+      ->AddVal("AgentId", id())
+      ->AddVal("Time", context()->time())
+      ->AddVal("Event", name)
+      ->AddVal("Value", val)
+      ->Record();
+}
+
+void Reactor::RecordStatus(std::string status, double power) {
+  context()
+      ->NewDatum("ReactorStatus")
+      ->AddVal("AgentId", id())
+      ->AddVal("Time", context()->time())
+      ->AddVal("Status", status)
+      ->AddVal("Power", power)
+      ->Record();
+}
+
+void Reactor::RecordInventories(double core, double reserve, double storage,
+                                double blanket, double helium) {
+  context()
+      ->NewDatum("ReactorInventories")
+      ->AddVal("AgentId", id())
+      ->AddVal("Time", context()->time())
+      ->AddVal("TritiumCore", core)
+      ->AddVal("TritiumReserve", reserve)
+      ->AddVal("TritiumStorage", storage)
+      ->AddVal("LithiumBlanket", blanket)
+      ->AddVal("HeliumStorage", helium)
+      ->Record();
+}
+
+void Reactor::DepleteBlanket(double bred_tritium_mass) {
+  cyclus::Material::Ptr blanket_mat = blanket.Pop();
+
+  cyclus::CompMap b = blanket_mat->comp()->mass();
+  cyclus::compmath::Normalize(&b, blanket_mat->quantity());
+
+  // Percent (as decimal) of T which comes from Li-7 instead of Li-6
+  double Li7_contribution = 0.15;
+
+  cyclus::CompMap depleted_comp;
+
+  // This is ALMOST the correct behavior, but "scraping the bottom of the
+  // barrel" is a little too complex for this implementation.
+  if ((b[30060000] - (1 - Li7_contribution) * 2 * bred_tritium_mass > 0) &&
```

Comment:
How should we treat the edge case (probably just a test case) of 100% Li-6 or Li7 (Li7= 1 or 0)? Maybe the && should be an or

Reply:
I think because of the inclusion of Li7_contribution in the equation, the && works (on the edges at 0 and 1 it will drop the unneeded term because of a multiplication by either Li7_contribution or 1-Li7_contribution), but I definitely think adding a test to make sure that the values of bred tritium are consistent is a good idea.

Reply:
Done.